File: src/flutter/flet_svg/lib/src/flet_svg.dart
import 'dart:io';
import 'dart:convert';
import 'dart:typed_data';
import 'package:flet/flet.dart';
import 'package:flutter/material.dart';
import 'package:flutter_svg/flutter_svg.dart';

class FletSvgControl extends StatelessWidget {
  final Control? parent;
  final Control control;

  const FletSvgControl({
    super.key,
    required this.parent,
    required this.control,
  });

  @override
  Widget build(BuildContext context) {
    String src = control.attrString('src') ?? '';
    String kind = control.attrString('kind') ?? 'network';
    String? key = control.attrString('key');
    String fitStr = control.attrString('fit') ?? 'contain';
    String alignmentStr = control.attrString('alignment') ?? 'center';
    double? width = control.attrDouble('width');
    double? height = control.attrDouble('height');
    bool matchTextDirection = control.attrBool('match_text_direction') ?? false;
    bool allowDrawingOutsideViewBox = control.attrBool('allow_drawing_outside_view_box') ?? false;
    String? semanticsLabel = control.attrString('semantics_label');
    bool excludeFromSemantics = control.attrBool('exclude_from_semantics') ?? false;
    String clipBehaviorStr = control.attrString('clip_behavior') ?? 'hard_edge';
    BoxFit fit = switch(fitStr) {
      'fill' => BoxFit.fill,
      'cover' => BoxFit.cover,
      'fit_width' => BoxFit.fitWidth,
      'fit_height' => BoxFit.fitHeight,
      'none' => BoxFit.none,
      'scale_down' => BoxFit.scaleDown,
      _ => BoxFit.contain,
    };
    AlignmentGeometry alignment = switch(alignmentStr) {
      'bottom_right' => Alignment.bottomRight,
      'bottom_center' => Alignment.bottomCenter,
      'bottom_left' => Alignment.bottomLeft,
      'top_right' => Alignment.topRight,
      'top_center' => Alignment.topCenter,
      'top_left' => Alignment.topLeft,
      _ => Alignment.center,
    };
    Clip clipBehavior = switch(clipBehaviorStr) {
      'none' => Clip.none,
      'anti_alias' => Clip.antiAlias,
      'anti_alias_with_save_layer' => Clip.antiAliasWithSaveLayer,
      _ => Clip.hardEdge,
    };

    Widget myControl = switch(kind) {
      'asset' => SvgPicture.asset(
        src,
        key: key != null ? Key(key) : null,
        width: width,
        height: height,
        fit: fit,
        alignment: alignment,
        matchTextDirection: matchTextDirection,
        allowDrawingOutsideViewBox: allowDrawingOutsideViewBox,
        semanticsLabel: semanticsLabel,
        excludeFromSemantics: excludeFromSemantics,
        clipBehavior: clipBehavior
      ),
      'file' => SvgPicture.file(
        File(src),
        key: key != null ? Key(key) : null,
        width: width,
        height: height,
        fit: fit,
        alignment: alignment,
        matchTextDirection: matchTextDirection,
        allowDrawingOutsideViewBox: allowDrawingOutsideViewBox,
        semanticsLabel: semanticsLabel,
        excludeFromSemantics: excludeFromSemantics,
        clipBehavior: clipBehavior
      ),
      'network' => SvgPicture.network(
        src,
        key: key != null ? Key(key) : null,
        width: width,
        height: height,
        fit: fit,
        alignment: alignment,
        matchTextDirection: matchTextDirection,
        allowDrawingOutsideViewBox: allowDrawingOutsideViewBox,
        semanticsLabel: semanticsLabel,
        excludeFromSemantics: excludeFromSemantics,
        clipBehavior: clipBehavior
      ),
      'memory' => SvgPicture.memory(
        Uint8List.fromList(utf8.encode(src)),
        key: key != null ? Key(key) : null,
        width: width,
        height: height,
        fit: fit,
        alignment: alignment,
        matchTextDirection: matchTextDirection,
        allowDrawingOutsideViewBox: allowDrawingOutsideViewBox,
        semanticsLabel: semanticsLabel,
        excludeFromSemantics: excludeFromSemantics,
        clipBehavior: clipBehavior
      ),
      _ => SvgPicture.string(
        src,
        key: key != null ? Key(key) : null,
        width: width,
        height: height,
        fit: fit,
        alignment: alignment,
        matchTextDirection: matchTextDirection,
        allowDrawingOutsideViewBox: allowDrawingOutsideViewBox,
        semanticsLabel: semanticsLabel,
        excludeFromSemantics: excludeFromSemantics,
        clipBehavior: clipBehavior
      ),
    };

    return constrainedControl(context, myControl, parent, control);
  }
}

File: src/flet_svg/flet_svg.py
from enum import Enum
from typing import Any, Optional

from flet.core.constrained_control import ConstrainedControl
from flet.core.control import OptionalNumber

class FletSvg(ConstrainedControl):
    """
    FletSvg Control description.
    """

    def __init__(
        self,
        #
        # Control
        #
        opacity: OptionalNumber = None,
        tooltip: Optional[str] = None,
        visible: Optional[bool] = None,
        data: Any = None,
        #
        # ConstrainedControl
        #
        left: OptionalNumber = None,
        top: OptionalNumber = None,
        right: OptionalNumber = None,
        bottom: OptionalNumber = None,
        #
        # FletSvg specific
        #
        src: Optional[str] = None,
        kind: Optional[str] = None,
        key: Optional[str] = None,
        fit: str = 'contain',
        alignment: str = 'center',
        width: OptionalNumber = None,
        height: OptionalNumber = None,
        match_text_direction: bool = False,
        allow_drawing_outside_view_box: bool = False,
        semantics_label: Optional[str] = None,
        exclude_from_semantics: bool = False,
        clip_behavior: str = 'hard_edge'
    ):
        ConstrainedControl.__init__(
            self,
            tooltip=tooltip,
            opacity=opacity,
            visible=visible,
            data=data,
            left=left,
            top=top,
            right=right,
            bottom=bottom,
        )

        self.src = src
        self.kind = kind
        self.key = key
        self.fit = fit
        self.alignment = alignment
        self.width = width
        self.height = height
        self.match_text_direction = match_text_direction
        self.allow_drawing_outside_view_box = allow_drawing_outside_view_box
        self.semantics_label = semantics_label
        self.exclude_from_semantics = exclude_from_semantics
        self.clip_behavior = clip_behavior

    def _get_control_name(self):
        return "flet_svg"

    # src
    @property
    def src(self):
        return self._get_attr("src")

    @src.setter
    def src(self, src):
        self._set_attr("src", src)

    # kind
    @property
    def kind(self):
        return self._get_attr("kind")

    @kind.setter
    def kind(self, kind):
        self._set_attr("kind", kind)

    # key
    @property
    def key(self):
        return self._get_attr("key")

    @key.setter
    def key(self, key):
        self._set_attr("key", key)

    # fit
    @property
    def fit(self):
        return self._get_attr("fit")

    @fit.setter
    def fit(self, fit):
        self._set_attr("fit", fit)

    # alignment
    @property
    def alignment(self):
        return self._get_attr("alignment")

    @alignment.setter
    def alignment(self, alignment):
        self._set_attr("alignment", alignment)

    # width
    @property
    def width(self):
        return self._get_attr("width")

    @width.setter
    def width(self, width):
        self._set_attr("width", width)

    # height
    @property
    def height(self):
        return self._get_attr("height")

    @height.setter
    def height(self, height):
        self._set_attr("height", height)

    # match_text_direction
    @property
    def match_text_direction(self):
        return self._get_attr("match_text_direction")

    @match_text_direction.setter
    def match_text_direction(self, match_text_direction):
        self._set_attr("match_text_direction", match_text_direction)

    # allow_drawing_outside_view_box
    @property
    def allow_drawing_outside_view_box(self):
        return self._get_attr("allow_drawing_outside_view_box")

    @allow_drawing_outside_view_box.setter
    def allow_drawing_outside_view_box(self, allow_drawing_outside_view_box):
        self._set_attr("allow_drawing_outside_view_box", allow_drawing_outside_view_box)

    # semantics_label
    @property
    def semantics_label(self):
        return self._get_attr("semantics_label")

    @semantics_label.setter
    def semantics_label(self, semantics_label):
        self._set_attr("semantics_label", semantics_label)

    # exclude_from_semantics
    @property
    def exclude_from_semantics(self):
        return self._get_attr("exclude_from_semantics")

    @exclude_from_semantics.setter
    def exclude_from_semantics(self, exclude_from_semantics):
        self._set_attr("exclude_from_semantics", exclude_from_semantics)

    # clip_behavior
    @property
    def clip_behavior(self):
        return self._get_attr("clip_behavior")

    @clip_behavior.setter
    def clip_behavior(self, clip_behavior):
        self._set_attr("clip_behavior", clip_behavior)

